src/native/aws_signing.c

```diff
@@ -0,0 +1,325 @@
+    jobject java_original_request;
+    jobject java_sign_param_predicate;
+    struct aws_http_message *native_request;
+    struct aws_signable *original_message_signable;
+    struct aws_string *region;
+    struct aws_string *service;
+};
+
+static void s_cleanup_callback_data(struct s_aws_sign_request_callback_data *callback_data) {
+
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    (*env)->DeleteGlobalRef(env, callback_data->java_future);
+    (*env)->DeleteGlobalRef(env, callback_data->java_original_request);
+
+    if (callback_data->java_sign_param_predicate) {
+        (*env)->DeleteGlobalRef(env, callback_data->java_sign_param_predicate);
+    }
+
+    if (callback_data->native_request) {
+        aws_http_message_release(callback_data->native_request);
+    }
+
+    if (callback_data->original_message_signable) {
+        aws_signable_destroy(callback_data->original_message_signable);
+    }
+
+    aws_string_destroy(callback_data->region);
+    aws_string_destroy(callback_data->service);
+
+    aws_mem_release(aws_jni_get_allocator(), callback_data);
+}
+
+static jobject s_create_signed_java_http_request(
+    JNIEnv *env,
+    struct aws_http_message *native_request,
+    jobject java_original_request) {
+    jstring jni_uri = NULL;
+    jobjectArray jni_headers = NULL;
+    jobject http_request = NULL;
+
+    jstring jni_method = (*env)->GetObjectField(env, java_original_request, http_request_properties.method_field_id);
+    jobject jni_body_stream =
+        (*env)->GetObjectField(env, java_original_request, http_request_properties.body_stream_field_id);
+
+    struct aws_byte_cursor path_cursor;
+    if (aws_http_message_get_request_path(native_request, &path_cursor)) {
+        goto done;
+    }
+
+    jni_uri = aws_jni_string_from_cursor(env, &path_cursor);
+    if (jni_uri == NULL) {
+        goto done;
+    }
+
+    jni_headers = aws_java_headers_array_from_http_headers(env, aws_http_message_get_headers(native_request));
+    if (jni_headers == NULL) {
+        goto done;
+    }
+
+    http_request = (*env)->NewObject(
+        env,
+        http_request_properties.http_request_class,
+        http_request_properties.constructor_method_id,
+        jni_method,
+        jni_uri,
+        jni_headers,
+        jni_body_stream);
+
+done:
+
+    if (jni_uri != NULL) {
+        (*env)->DeleteLocalRef(env, jni_uri);
+    }
+
+    if (jni_headers != NULL) {
+        (*env)->DeleteLocalRef(env, jni_headers);
+    }
+
+    return http_request;
+}
+
+static void s_complete_signing_exceptionally(
+    JNIEnv *env,
+    struct s_aws_sign_request_callback_data *callback_data,
+    int error_code) {
+
+    jint jni_error_code = error_code;
+    struct aws_byte_cursor error_cursor = aws_byte_cursor_from_c_str(aws_error_name(error_code));
+    jstring jni_error_string = aws_jni_string_from_cursor(env, &error_cursor);
+    AWS_FATAL_ASSERT(jni_error_string);
+
+    jobject crt_exception = (*env)->NewObject(
```

Comment:
do we need to delete the returned reference?

Reply:
Yes